New folder linux

Coder206 opened the pull request:
<!-- Please describe your changes on the following line: -->

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #11983 (github issue number if applicable).
- [x] There are tests for these changes (./mach package)

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12850)

<!-- Reviewable:end -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @simonsapin (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/servo/package_commands.py

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify gfx code, but no tests are modified. Please consider adding a test!

emilio commented:
r? @aneeshusa

Coder206 commented:
@ConnorGBrewster Is this what you were looking expecting?

ConnorGBrewster commented:
Yes, this is what I had in mind. Although I will let @aneeshusa review this still, since I don't know all of our python best practices.

Coder206 commented:
@ConnorGBrewster Perfect! That makes sense. :)

ConnorGBrewster commented:
No worries! Thanks for doing it! 😄

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -209,24 +209,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
         else:
+            dir_to_temp = '/'.join(binary_path.split('/')[:-2]) + '/targz'
             dir_to_package = '/'.join(binary_path.split('/')[:-1])
             dir_to_root = '/'.join(binary_path.split('/')[:-3])
```
dir_to_root is no longer used.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -209,24 +209,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             browserhtml_path = find_dep_path_newest('browserhtml', binary_path)
             if browserhtml_path is None:
                 print("Could not find browserhtml package; perhaps you haven't built Servo.")
                 return 1
```
Please move the 'find browserhtml' section before the 'delete old `dir_to_temp`' section.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -209,24 +209,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             msi_path = path.join(dir_to_msi, "Servo.msi")
             print("Packaged Servo into {}".format(msi_path))
         else:
+            dir_to_temp = '/'.join(binary_path.split('/')[:-2]) + '/targz'
```
Can you please replace all of the `.split('/')` and `'/'.join()` calls with `path.dirname` and `path.join`? All this string processing hurts my eyes :(

Coder206 replied:
@aneeshusa What does `[:-2]` do?

aneeshusa replied:
It's the Python slice operator. In a Python2 REPL:
```
[aneesh@cheetah servo]$ python2                                                                       [18:59:10]
Python 2.7.12 (default, Jun 28 2016, 08:31:05)
[GCC 6.1.1 20160602] on linux2
Type help, copyright, credits or license for more information.
>>> binary_path = '/home/aneesh/workspace/servo'
>>> binary_path.split('/')
['', 'home', 'aneesh', 'workspace', 'servo']
>>> binary_path.split('/')[:-2]
['', 'home', 'aneesh']
>>> [only_one_element_here][:-2]
[]
```

Essentially, it takes a list and returns another list with the last two elements removed. Here it's being used to strip off two path components from `binary_path`.

Coder206 replied:
Thank you @aneeshusa!

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -235,17 +232,20 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             os.close(runservo)
+
             print("Creating tarball")
             tar_path = '/'.join(dir_to_package.split('/')[:-1]) + '/'
```
This can be just `tar_path = self.get_target_dir()`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -235,17 +232,20 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             tar_path = '/'.join(dir_to_package.split('/')[:-1]) + '/'
             time = datetime.utcnow().replace(microsecond=0).isoformat()
             time = time.replace(':', "-")
             tar_path += time + "-servo-tech-demo.tar.gz"
```
Use `tar_path = path.join(tar_path, time + ...)` instead of `+=` here.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -209,24 +209,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-                if 'browserhtml' not in f:
-                    delete(dir_to_package + '/build/' + f)
+            print("Copying files")
+            shutil.copytree(dir_to_package + '/resources', resources_dir)
```
We need to copy this from `path.join(self.get_top_dir(), 'resources'` instead of `dir_to_package`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -209,24 +209,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+            print("Copying files")
+            shutil.copytree(dir_to_package + '/resources', resources_dir)
+            shutil.copytree(browserhtml_path, resources_dir + browserhtml_path.split('/')[-1])
+
```
We also need to copy the actual `servo` binary!

Coder206 replied:
@aneeshusa I failed to copy the binary eith this: `shutil.copytree(path.join(dir_to_temp[:-5], 'servo'), dir_to_temp)`

aneeshusa replied:
The path to the servo binary is just `binary_path` :)

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -209,24 +209,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-                    delete(dir_to_package + '/build/' + f)
+            print("Copying files")
+            shutil.copytree(dir_to_package + '/resources', resources_dir)
+            shutil.copytree(browserhtml_path, resources_dir + browserhtml_path.split('/')[-1])
```
I believe browserhtml needs to go under a `build` directory - another place to use `path.join`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -209,24 +209,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             print("Packaged Servo into {}".format(msi_path))
         else:
+            dir_to_temp = '/'.join(binary_path.split('/')[:-2]) + '/targz'
             dir_to_package = '/'.join(binary_path.split('/')[:-1])
```
We can get rid of `dir_to_package`.

aneeshusa commented:
Sorry about the delay in reviewing! Please also try to run Servo from the package tarball to make sure the packaging works properly.

aneeshusa commented:
@Coder206 How is this going? Do you have any questions?

Coder206 commented on python/servo/package_commands.py:
```diff
@@ -209,24 +209,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+            shutil.copytree(dir_to_package + '/resources', resources_dir)
+            shutil.copytree(browserhtml_path, resources_dir + browserhtml_path.split('/')[-1])
+
             print("Writing runservo.sh")
```
@aneeshusa Is it still necessary to make a runservo.sh here?

aneeshusa replied:
@ConnorGBrewster has a PR to remove it: #12154.
It looks like it would be pretty easy to fold that PR into this one.

Coder206 replied:
So should I work to combine the two PRs?

aneeshusa replied:
Actually, I just realized that removing `runservo.sh` will require some additional changes, so let's keep `runservo.sh` for now and handle the removal in the other PR.

Coder206 committed:
Use a temp folder when packaging on Linux

The purpose of the code is to make a new temporary folder with the
necessary resources used to package Linux. Before this, the package
was built based on a modified target directory. By using a temporary
folder, it removes the need to rebuild Servo every time it gets
packaged for Linux.

Coder206 commented:
@aneeshusa Squash?

aneeshusa commented:
@Coder206 no need to squash quite yet. Looking better but I still have some review comments.

- Please review the [`os.path.join` documentation](https://docs.python.org/2/library/os.path.html#os.path.join), which we've aliased as just `path.join` in this file. Essentially, each call to `path.join` should look like this:
```
>>> top_dir = "/zoo"
>>> path.join(top_dir, 'cow')
'/zoo/cow'
>>> path.join(top_dir, 'cow', 'say')
'/zoo/cow/say'
```
Note that each call to `path.join` has two or more arguments. In many cases you are passing one argument, or doing something like `path.join(var_a + var_b)`, which should be `path.join(var_a, var_b)`. The goal is to not be using `+` on these strings representing paths.
Please review your uses of `path.join` and update accordingly; let me know if you have questions.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,43 +287,42 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+
             print("Creating tarball")
-            tar_path = '/'.join(dir_to_package.split('/')[:-1]) + '/'
+            tar_path = self.get_target_dir()
```
Just inline this expression into the `path.join` call below.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,43 +287,42 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            for f in os.listdir(dir_to_package + '/build/'):
-                if 'browserhtml' not in f:
-                    delete(dir_to_package + '/build/' + f)
+            if os.path.exists(dir_to_temp):
```
This can be just `path.exists` since `os.path` is aliased as `path`.

aneeshusa replied:
Also, please add this comment above this line:
```
# TODO(aneeshusa): lock dir_to_temp to prevent simultaneous builds
```

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,43 +287,42 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+                print("Cleaning up from previous packaging")
+                delete(dir_to_temp)
+            print("Copying files")
+            shutil.copytree(path.join(self.get_top_dir(), 'resources'), resources_dir)
```
This is a good example of how to use `path.join` correctly.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,43 +287,42 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             print("Writing runservo.sh")
             # TODO: deduplicate this arg list from post_build_commands
-            servo_args = ['-w', '-b',
+            servo_args = ['-w', '-b', '-M', '-S',
```
Please leave this unchanged in this PR.

Coder206 commented on python/servo/package_commands.py:
```diff
@@ -287,43 +287,42 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+            print("Copying files")
+            shutil.copytree(path.join(self.get_top_dir(), 'resources'), resources_dir)
+            shutil.copytree(browserhtml_path, path.join(dir_to_temp + "/build"))
+            shutil.copytree(path.join(dir_to_temp[:-5], 'servo'), dir_to_temp)
```
@aneeshusa I am getting this error here:
```

OSError: [Errno 20] Not a directory: '/home/user/servo/target/debug/servo'

  File "/home/user/servo/python/servo/package_commands.py", line 306, in package
    shutil.copytree(path.join(dir_to_temp[:-5], 'servo'), dir_to_temp)
  File "/usr/lib/python2.7/shutil.py", line 171, in copytree
    names = os.listdir(src)
```

aneeshusa replied:
I think may have missed my [previous comment](https://github.com/servo/servo/pull/12850#discussion_r78609350), but `binary_path` already has the path to the servo binary. Also, looking at the [docs for the `shutil` module](https://docs.python.org/2/library/shutil.html), you'll see that `shutil.copytree` is for copying directories recursively, not files. Since the servo binary is just one file, another method from `shutil` would be more appropriate here.

Coder206 commented:
@aneeshusa Here is an updated version reflecting your most recent changes: https://gist.github.com/Coder206/66d659701610cb20508db6fa92cf49e5

aneeshusa commented:
@Coder206 instead of a gist, please continue making new commits on top of this branch and pushing them up to make it possible to view diffs and leave comments inline. (We can always squash or rearrange the commits later.)

Coder206 commented:
@aneeshusa Noted. Sorry about that, in hindsight it makes sense!

Coder206 commented on python/servo/package_commands.py:
```diff
@@ -287,24 +287,24 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-                delete(resources_dir)
-            shutil.copytree(dir_to_root + '/resources', resources_dir)
+            dir_to_temp = path.join(binary_path[:-5], 'targz')
+            resources_dir = path.join(dir_to_temp, 'resources')
```
@aneeshusa Can I rename this to "dir_to_resources"?

Coder206 commented on python/servo/package_commands.py:
```diff
@@ -287,24 +287,24 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+            dir_to_temp = path.join(binary_path[:-5], 'targz')
+            resources_dir = path.join(dir_to_temp, 'resources')
             browserhtml_path = find_dep_path_newest('browserhtml', binary_path)
+            
```
@aneeshua Looks like the tidy failed because of a few spaces like here.

KiChjang commented:
@aneeshusa Reviewer ping!

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,24 +287,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            if os.path.exists(resources_dir):
-                delete(resources_dir)
-            shutil.copytree(dir_to_root + '/resources', resources_dir)
+            dir_to_temp = path.join(binary_path[:-5], 'targz')
```
Use `os.path.dirname` instead of `binary_path[:-5]`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,24 +287,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            if os.path.exists(resources_dir):
-                delete(resources_dir)
-            shutil.copytree(dir_to_root + '/resources', resources_dir)
+            dir_to_temp = path.join(os.path.dirname(binary_path), 'targz')
```
We can just use `path.dirname` instead of `os.path.dirname`.

aneeshusa replied:
Also, instead of `targz`, let's use a more descriptive name like `packaging-temp`. This is also packaging-technology neutral, which will help us possibly combine these blocks down the road.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,24 +287,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-                delete(resources_dir)
-            shutil.copytree(dir_to_root + '/resources', resources_dir)
+            dir_to_temp = path.join(os.path.dirname(binary_path), 'targz')
+            resources_dir = path.join(dir_to_temp, 'resources')
```
I think it's more clear to move this line to right before the line where we use it, as it's not used anywhere else.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,24 +287,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+                delete(dir_to_temp)
+            print("Copying files")
+            shutil.copytree(path.join(self.get_top_dir(), 'resources'), resources_dir)
+            shutil.copytree(browserhtml_path, path.join(dir_to_temp, "build"))
```
nit: use single quotes for `build`

aneeshusa reviewed (changes requested):
This looks pretty good now, just a few nits left. Go ahead and squash when you address these.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,24 +287,21 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+                # TODO(aneeshusa): lock dir_to_temp to prevent simultaneous builds
+                print("Cleaning up from previous packaging")
+                delete(dir_to_temp)
+            print("Copying files")
```
nit: add an empty line before this one

aneeshusa commented:
Please update the commit message as well to reflect what you did and the rationale behind it.

aneeshusa commented:
@Coder206 The commit title looks great now, but the commit message body could still use some work.  The body is generally used to mention the why behind the changes, point out any tricky, etc. changes that deserve special attention, or provide more description than fits within the title. Some concrete suggestions:

- No need to name authors (git tracks this), reviewers or requesters (github tracks this)
- The rest of the body right now describes the changes (which is good); see if you make that description shorter. (If the description gets too long, it becomes no faster than just reading the diff directly.)
- Add an explanation of why this change was made: what are the benefits of using a temporary directory instead of modifying the target dir itself?

aneeshusa commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit be87452 has been approved by `aneeshusa`

<!-- @bors-servo r=aneeshusa be874525dd530a9a5fd781dac8fab90ed65ed394 -->

bors-servo commented:
:hourglass: Testing commit be87452 with merge 968d4f0...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3338)

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -287,24 +287,22 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
+                # TODO(aneeshusa): lock dir_to_temp to prevent simultaneous builds
+                print("Cleaning up from previous packaging")
+                delete(dir_to_temp)
+                
```
This line has whitespace, remove it (line should just be empty).

aneeshusa reviewed (changes requested):
Always remember to run `./mach test-tidy` before pushing :)

KiChjang commented:
@bors-servo r=aneeshusa

bors-servo commented:
:pushpin: Commit d672750 has been approved by `aneeshusa`

<!-- @bors-servo r=aneeshusa d672750ca4f702857466728a38831fe57a955d52 -->

bors-servo commented:
:hourglass: Testing commit d672750 with merge 7fbd35e...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3252), [arm64](http://build.servo.org/builders/arm64/builds/3265), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3340), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3520), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3329), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3309), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1215)

Coder206 commented:
Thank you very much @aneeshusa!
